packages/pages/src/common/src/template/hydration.ts

```diff
@@ -0,0 +1,188 @@
+import { HeadConfig, renderHeadConfigToString } from "./head.js";
+import { convertToPosixPath } from "./paths.js";
+import { TemplateRenderProps } from "./types.js";
+
+/**
+ * Imports the custom hydration template and entrypoint template as modules and calls
+ * the render function.
+ *
+ * @param clientRenderTemplatePath the path to the custom client render template
+ * @param templateModulePath the path to the template module
+ * @param props the {@link TemplateRenderProps}
+ * @returns the HTML as a string
+ */
+export const getHydrationTemplate = (
+  clientRenderTemplatePath: string,
+  templateModulePath: string,
+  props: TemplateRenderProps
+): string => {
+  return `
+        import {default as Component} from "${convertToPosixPath(
+          templateModulePath
+        )}";
+        import {render} from "${convertToPosixPath(clientRenderTemplatePath)}";
+        render(
+        {
+            Page: Component,
+            pageProps: ${JSON.stringify(props)},
+        }
+        );
+    `;
+};
+
+/**
+ * Get the server template with injected html common to both the dev and plugin side of things.
+ * For the most part, injects data into the <head> tag. It also provides validation.
+ *
+ * @param clientHtml
+ * @param serverHtml
+ * @param appLanguage
+ * @param headConfig
+ * @returns the server template with injected html
+ */
+const getCommonInjectedServerHtml = (
+  clientHtml: string,
+  serverHtml: string,
+  appLanguage: string,
+  headConfig?: HeadConfig
+): string => {
+  validateHeadTagExists(serverHtml);
+
+  // Add the language to the <html> tag if it exists
+  serverHtml.replace("<!--app-lang-->", appLanguage);
+
+  serverHtml = injectIntoHead(
+    serverHtml,
+    `<script type="module">${clientHtml}</script>`
```

Comment:
also why would the client and server html intermingle? They should be separate right?

Reply:
Gotcha, that makes sense. Before I think it was just a separately-bundled file but now we're including it's dependencies in the server bundle. Any concerns about that increasing the server-size?

Reply:
I'm not 100% clear what you mean. We're effectively doing the same thing as before. Prior to this change the wrapper.ts used in the plugin did this:

```
return `<!DOCTYPE html>
    <html lang=${lang}>
    <head>
        <script>window.__INITIAL__DATA__ = ${JSON.stringify(props)}</script>
        // css stuff
        // head config stuff
    </head>
    <body>
        <div id="reactele">${template}</div>${
    hydrate
      ? `<script type="module" src="${
          props.relativePrefixToRoot +
          findHydrationFilename(
            `${projectFilepaths.hydrationBundleOutputRoot}/${templateModuleInternal.templateName}.tsx`,
            props
          )
        }" defer></script>`
      : ""
  }
    </body>
    </html>`;
};
```

The only difference now is that we use the _server.tsx and _client.tsx from the dist/assets folder. The templates in dist/assets/server are the same where they are just the bundled template part. The things in hydration.ts just put it all together. The effect is the same:
```
// read _server.tsx with dist/assets/server stringified code in it (same as above <div id="reactele">${template}</div>)
// inject below into its <script type="module>HERE</script> for hydration (also same as above hydrate? part)

import {default as Component} from "pathToTemplateModule...aka the file in dist/assets/server";
import {render} from "_client.tsx";
render(
{
    Page: Component, // this is the dist/assets/server template again
    pageProps: ${JSON.stringify(props)},
}
);
```
The total size of what's returned here should be the same. The added benefit actually is that there are no longer compiled hydration templates required in the plugin AND we don't need to add a props json on the page to read from.

I guess maybe what you're saying is that in the script module we're dumping the string rather than having it be a referenced file that's separate?

Also did you want me to do this?
> I can change it from clientHtml to clientRenderString if you think that's better.

Reply:
> I guess maybe what you're saying is that in the script module we're dumping the string rather than having it be a referenced file that's separate?

Yeah I was thinking it might be better to not inline the JS there. It could cause issues on some sites which want to restrict it.

> Also did you want me to do this?

Yeah that makes it more clear to me.